### PR TITLE
fix(XamlGen): Handle base type not specified in code-behind

### LINKED ISSUE #7404: Unable to use binding on XAML files with code-behind missing a base type
### Current behavior

```xml
<views:StatusView DataContext="{Binding StatusViewModel}" />
```
WASM(maybe other) show this error:
```
error CS1061: 'DependencyObject' does not contain a definition for 'SetBinding' and no accessible extension method 'SetBinding' accepting a first argument of type 'DependencyObject' could be found (are you missing a using directive or an assembly reference?)
```

### Expected behavior

_No response_

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

3.12.0-dev.2

### Affected platforms

WebAssembly

### IDE

Visual Studio 2022

### IDE version

preview 7.0

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #7404

## PR Type

What kind of change does this PR introduce?

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SourceGenerators/Uno.UI.SourceGenerators.net6.Tests/Windows_UI_Xaml_Data/BindingTests/Given_Binding.cs [hunk 1]
```diff
@@ -33,4 +34,86 @@ public async Task When_Binding_ElementName_In_Template()
...
+							}
+						}
+					}
+					"""
+				}
+			}
+		};
+		test.ExpectedDiagnostics.Add(
+			// /0/Test0.cs(3,30): warning UXAML0002: TestRepro.UserControl1 does not explicitly define the Windows.UI.Xaml.Controls.UserControl base type in code behind.
+			DiagnosticResult.CompilerWarning("UXAML0002").WithSpan(3, 30, 3, 42).WithArguments("TestRepro.UserControl1 does not explicitly define the Windows.UI.Xaml.Controls.UserControl base type in code behind.")
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators.net6.Tests/Windows_UI_Xaml_Data/BindingTests/Given_Binding.cs:115]: @jeromelaban Is this warning still relevant?

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs [hunk 2]
```diff
@@ -288,6 +288,23 @@ private string GetSourceLink(Uno.Roslyn.MSBuildItem projectItemInstance)
...
+					{
+						continue;
+					}
+
+					var xClassSymbol = XamlFileGenerator.FindClassSymbol(topLevelControl, _metadataHelper);
+					if (xClassSymbol is not null)
+					{
+						xamlTypeToXamlTypeBaseMap.Add(xClassSymbol, topLevelControl.Type);
+					}
+				}
```
Comment (Youssef1313) [inline src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:306]: We need to make sure this doesn't hurt the performance. The reported issue is quite a niche case where we already produce a warning to guide the user towards an easy workaround (specify base type in codebehind).

If this hurts performance badly, I personally don't think we should take this change.

**Comment by amelield** [comment]
So if the base type is not specified, it would deduce it with the information present in the file? This is cool.

**Comment by Youssef1313** [comment]
> So if the base type is not specified, it would deduce it with the information present in the file? This is cool.

Yes. The idea is that previously we are able only to get the base type from code-behind (C# sources). Now we have a pre-generation pass that collects base types from XAML sources.

**Comment by Youssef1313** [comment]
@jeromelaban This is affecting the performance very badly, unfortunately.

File: src/SourceGenerators/Uno.UI.SourceGenerators.net6.Tests/Windows_UI_Xaml_Data/BindingTests/Given_Binding.cs (hunk 1 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators.net6.Tests/Windows_UI_Xaml_Data/BindingTests/Given_Binding.cs:115]: It should not be used anymore, indeed.

**Comment by jeromelaban** [comment]
> @jeromelaban This is affecting the performance very badly, unfortunately.

By how much?

**Comment by Youssef1313** [comment]
> > @jeromelaban This is affecting the performance very badly, unfortunately.
> 
> By how much?

I'm not sure I'll be able to say exact numbers, but this is what I see in binlog.

image

Though, the numbers in binary log don't consider concurrent execution, but it gives an insight that this is making performance somewhat noticeable worse.

Let me know if you have a better way for measuring.

File: src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs (hunk 2 above)
Comment (jeromelaban) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs:306]: Maybe we should be doing this in parallel, as there are many files to process.

Comment (Youssef1313) [reply src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlCodeGeneration.cs]: @jeromelaban Thanks! This has made a significant difference.

**Comment by Youssef1313** [comment]
From what I see in binlog, it's almost identical.

**Comment by jeromelaban** [comment]
Very nice! Looks like it'll be good to go then.

**Review by jeromelaban — APPROVED** [review]

**Pull request merged**